File: src/lib/grooveData.ts
/**
 * Core data model for a drum groove.
 *
 * The URL is the document: everything in GrooveData round-trips through the
 * query string (see urlCodec.ts). Lanes are flat arrays of cells covering all
 * measures; a cell is `null` (rest) or a lane-specific hit type.
 */

export interface TimeSig {
  top: number;
  bottom: number;
}

export type HihatHit =
  | 'normal'
  | 'accent'
  | 'open'
  | 'close'
  | 'ride'
  | 'rideBell'
  | 'crash'
  | 'stacker'
  | 'cowbell'
  | 'metronomeNormal'
  | 'metronomeAccent';

export type SnareHit =
  | 'normal'
  | 'accent'
  | 'ghost'
  | 'xstick'
  | 'flam'
  | 'drag'
  | 'buzz';

export type KickHit = 'normal' | 'splash' | 'kickAndSplash';

export type TomHit = 'normal';

export type Cell<H> = H | null;

export interface GrooveData {
  timeSig: TimeSig;
  /** Grid resolution: cells per whole note. 16 = 16ths, 12 = 8th triplets… */
  div: number;
  tempo: number;
  measures: number;
  /** Swing percentage 0–100 (straight divisions only). */
  swing: number;
  /** Metronome click density: 0 = off, 4 = quarters, 8 = eighths, 16 = sixteenths. */
  metronomeFreq: number;
  title: string;
  author: string;
  comments: string;
  hihat: Cell<HihatHit>[];
  snare: Cell<SnareHit>[];
  kick: Cell<KickHit>[];
  /** Four tom lanes (T1–T4), each same length as the other lanes. */
  toms: [Cell<TomHit>[], Cell<TomHit>[], Cell<TomHit>[], Cell<TomHit>[]];
}

export const ALLOWED_DIVS = [8, 12, 16, 24, 32, 48] as const;
export const ALLOWED_BOTTOMS = [2, 4, 8, 16] as const;
export const ALLOWED_METRONOME_FREQS = [0, 4, 8, 16] as const;

export const DEFAULT_TEMPO = 80;
export const MAX_MEASURES = 100;

/** Triplet grids are the divisions divisible by 12 (12, 24, 48). */
export function isTripletDiv(div: number): boolean {
  return div % 12 === 0;
}

/** Cells in one measure: div is per-whole-note, so scale by the time signature. */
export function cellsPerMeasure(timeSig: TimeSig, div: number): number {
  return Math.round((div * timeSig.top) / timeSig.bottom);
}

/** Cells per beat (one beat = 1/bottom note). */
export function cellsPerBeat(timeSig: TimeSig, div: number): number {
  return Math.round(div / timeSig.bottom);
}

export function totalCells(g: Pick<GrooveData, 'timeSig' | 'div' | 'measures'>): number {
  return cellsPerMeasure(g.timeSig, g.div) * g.measures;
}

export function emptyLane<H>(length: number): Cell<H>[] {
  return new Array<Cell<H>>(length).fill(null);
}

export function createEmptyGroove(overrides: Partial<GrooveData> = {}): GrooveData {
  const base: GrooveData = {
    timeSig: { top: 4, bottom: 4 },
    div: 16,
    tempo: DEFAULT_TEMPO,
    measures: 1,
    swing: 0,
    metronomeFreq: 0,
    title: '',
    author: '',
    comments: '',
    hihat: [],
    snare: [],
    kick: [],
    toms: [[], [], [], []],
  };
  const merged = { ...base, ...overrides };
  const n = totalCells(merged);
  merged.hihat = overrides.hihat ?? emptyLane(n);
  merged.snare = overrides.snare ?? emptyLane(n);
  merged.kick = overrides.kick ?? emptyLane(n);
  merged.toms = overrides.toms ?? [emptyLane(n), emptyLane(n), emptyLane(n), emptyLane(n)];
  return merged;
}

/**
 * Remap a lane onto a new grid size, preserving musical position as closely
 * as possible (exact for integer up/down-scaling; first-wins on collisions).
 */
export function remapLane<H>(lane: Cell<H>[], newLength: number): Cell<H>[] {
  const out = emptyLane<H>(newLength);
  if (lane.length === 0 || newLength === 0) return out;
  for (let i = 0; i < lane.length; i++) {
    const hit = lane[i];
    if (hit === null) continue;
    const j = Math.floor((i * newLength) / lane.length);
    if (out[j] === null) out[j] = hit;
  }
  return out;
}

/* ------------------------------------------------------------------ */
/* Tab-character maps (the URL contract — context-sensitive per lane) */
/* ------------------------------------------------------------------ */

export const HIHAT_CHAR_TO_HIT: Record<string, HihatHit> = {
  x: 'normal',
  X: 'accent',
  o: 'open',
  '+': 'close',
  r: 'ride',
  R: 'ride',
  b: 'rideBell',
  B: 'rideBell',
  c: 'crash',
  s: 'stacker',
  m: 'cowbell',
  n: 'metronomeNormal',
  N: 'metronomeAccent',
};

export const HIHAT_HIT_TO_CHAR: Record<HihatHit, string> = {
  normal: 'x',
  accent: 'X',
  open: 'o',
  close: '+',
  ride: 'r',
  rideBell: 'b',
  crash: 'c',
  stacker: 's',
  cowbell: 'm',
  metronomeNormal: 'n',
  metronomeAccent: 'N',
};

export const SNARE_CHAR_TO_HIT: Record<string, SnareHit> = {
  o: 'normal',
  O: 'accent',
  g: 'ghost',
  x: 'xstick',
  f: 'flam',
  d: 'drag',
  b: 'buzz',
  B: 'buzz',
};

export const SNARE_HIT_TO_CHAR: Record<SnareHit, string> = {
  normal: 'o',
  accent: 'O',
  ghost: 'g',
  xstick: 'x',
  flam: 'f',
  drag: 'd',
  buzz: 'b',
};

export const KICK_CHAR_TO_HIT: Record<string, KickHit> = {
  o: 'normal',
  x: 'splash',
  X: 'kickAndSplash',
};

export const KICK_HIT_TO_CHAR: Record<KickHit, string> = {
  normal: 'o',
  splash: 'x',
  kickAndSplash: 'X',
};

export const TOM_CHAR_TO_HIT: Record<string, TomHit> = {
  o: 'normal',
  x: 'normal',
};

export const TOM_HIT_TO_CHAR: Record<TomHit, string> = {
  normal: 'o',
};

File: ios/GrooveCore/Sources/GrooveCore/GrooveData.swift
import Foundation

// MARK: - Types
//
// Native Swift port of the web app's src/lib/grooveData.ts. Field names,
// validation clamps, and character maps are kept identical so grooves
// round-trip byte-for-byte with the website (see URLCodec).

public struct TimeSig: Equatable, Hashable, Codable, Sendable {
    public var top: Int
    public var bottom: Int
    public init(top: Int = 4, bottom: Int = 4) {
        self.top = top
        self.bottom = bottom
    }
}

public enum HihatHit: String, CaseIterable, Codable, Sendable {
    case normal, accent, open, close, ride, rideBell, crash, stacker, cowbell,
         metronomeNormal, metronomeAccent
}

public enum SnareHit: String, CaseIterable, Codable, Sendable {
    case normal, accent, ghost, xstick, flam, drag, buzz
}

public enum KickHit: String, CaseIterable, Codable, Sendable {
    case normal, splash, kickAndSplash
}

public enum TomHit: String, CaseIterable, Codable, Sendable {
    case normal
}

public enum Sticking: String, CaseIterable, Codable, Sendable {
    case R, L, B
}

public struct GrooveData: Equatable, Sendable {
    public var timeSig = TimeSig(top: 4, bottom: 4)
    /// Grid resolution: cells per whole note (16 = 16ths, 12 = 8th triplets…).
    public var div = 16
    public var tempo = GrooveConstants.defaultTempo
    public var measures = 1
    public var swing = 0
    public var metronomeFreq = 0
    public var title = ""
    public var author = ""
    public var comments = ""
    public var hihat: [HihatHit?] = []
    public var snare: [SnareHit?] = []
    public var kick: [KickHit?] = []
    /// Four tom lanes (T1…T4), each the same length as the other lanes.
    public var toms: [[TomHit?]] = [[], [], [], []]
    public var stickings: [Sticking?] = []

    public init() {}
}

// MARK: - Constants

public enum GrooveConstants {
    public static let allowedDivs = [8, 12, 16, 24, 32, 48]
    public static let allowedBottoms = [2, 4, 8, 16]
    public static let allowedMetronomeFreqs = [0, 4, 8, 16]
    public static let defaultTempo = 80
    public static let maxMeasures = 100
}

// MARK: - Grid math

/// Triplet grids are the divisions divisible by 12 (12, 24, 48).
public func isTripletDiv(_ div: Int) -> Bool { div % 12 == 0 }

/// Cells in one measure: div is per-whole-note, scaled by the time signature.
public func cellsPerMeasure(_ timeSig: TimeSig, _ div: Int) -> Int {
    Int((Double(div * timeSig.top) / Double(timeSig.bottom)).rounded())
}

/// Cells per beat (one beat = 1/bottom note).
public func cellsPerBeat(_ timeSig: TimeSig, _ div: Int) -> Int {
    Int((Double(div) / Double(timeSig.bottom)).rounded())
}

public func totalCells(_ g: GrooveData) -> Int {
    cellsPerMeasure(g.timeSig, g.div) * g.measures
}

public func emptyLane<T>(_ length: Int) -> [T?] {
    Array<T?>(repeating: nil, count: max(0, length))
}

public func laneHasHits<T>(_ lane: [T?]) -> Bool {
    lane.contains { $0 != nil }
}

/// Remap a lane onto a new grid size, preserving musical position as closely
/// as possible (exact for integer up/down-scaling; first-wins on collisions).
public func remapLane<T>(_ lane: [T?], _ newLength: Int) -> [T?] {
    var out: [T?] = emptyLane(newLength)
    if lane.isEmpty || newLength == 0 { return out }
    for i in 0..<lane.count {
        guard let hit = lane[i] else { continue }
        let j = (i * newLength) / lane.count
        if out[j] == nil { out[j] = hit }
    }
    return out
}

// MARK: - Character maps (the URL contract — context-sensitive per lane)

public enum TabMaps {
    public static let hihatCharToHit: [Character: HihatHit] = [
        "x": .normal, "X": .accent, "o": .open, "+": .close,
        "r": .ride, "R": .ride, "b": .rideBell, "B": .rideBell,
        "c": .crash, "s": .stacker, "m": .cowbell,
        "n": .metronomeNormal, "N": .metronomeAccent,
    ]
    public static let hihatHitToChar: [HihatHit: Character] = [
        .normal: "x", .accent: "X", .open: "o", .close: "+",
        .ride: "r", .rideBell: "b", .crash: "c", .stacker: "s",
        .cowbell: "m", .metronomeNormal: "n", .metronomeAccent: "N",
    ]

    public static let snareCharToHit: [Character: SnareHit] = [
        "o": .normal, "O": .accent, "g": .ghost, "x": .xstick,
        "f": .flam, "d": .drag, "b": .buzz, "B": .buzz,
    ]
    public static let snareHitToChar: [SnareHit: Character] = [
        .normal: "o", .accent: "O", .ghost: "g", .xstick: "x",
        .flam: "f", .drag: "d", .buzz: "b",
    ]

    public static let kickCharToHit: [Character: KickHit] = [
        "o": .normal, "x": .splash, "X": .kickAndSplash,
    ]
    public static let kickHitToChar: [KickHit: Character] = [
        .normal: "o", .splash: "x", .kickAndSplash: "X",
    ]

    public static let tomCharToHit: [Character: TomHit] = [
        "o": .normal, "x": .normal,
    ]
    public static let tomHitToChar: [TomHit: Character] = [
        .normal: "o",
    ]

    public static let stickingCharToHit: [Character: Sticking] = [
        "R": .R, "r": .R, "L": .L, "l": .L, "B": .B, "b": .B,
    ]
    public static let stickingHitToChar: [Sticking: Character] = [
        .R: "R", .L: "L", .B: "B",
    ]
}
